### ISSUE TITLE
1.5 LaTeX: xref inside desc_name not allowed

### ISSUE THREAD

**Issue by birkenfeld** [sphinx-doc/sphinx#3200]
In a custom extension, we're using x-refs inside `desc_name`, which generates LaTeX like this:
```
\pysigline{\sphinxbfcode{{\hyperref[deviceapi:nicos.core.device.Device.description]{\sphinxcrossref{description}}}}\sphinxstrong{ (a string, settable at runtime)}}
```

When running, it errors out with 
```
! Argument of \T1\. has an extra }.
```

Any ideas? Thanks!

**Comment by jfbu** [comment]
This issue is caused by `\hyperref` not being happy that`\sphinxbfcode` makes the ``.`` "active". There are other characters made active for reasons of the verbatim like character of `\sphinxcode/\sphinxbfcode` and also allow line-breaks. I started inserting these changes initially to get straight quotes in inline literals. I did not anticipage that ``\sphinxcode/bfcode`` would be used for other things that verbatim content.

If the extension can generate code of the following type:

    \hyperref[\detokenize{deviceapi:nicos.core.device.Device.description}]{...

this would fix the issue (with a working hyperlink.)

How much can the extension customize LaTeX output? 

Because there are a range of possibilities going from simply redefining ``\sphinxcode`` to just be ``\texttt`` like before I started fiddling with it, to customize the lists of characters made active, which are found in 

    \def\sphinx@literal@nolig@list {\do\`\do\<\do\>\do\'\do\-}
    \newcommand*\sphinxbreaksafteractivelist  {\do\.\do\,\do\;\do\?\do\!\do\/}

Of course one does not want to alter code-blocks. The first list above is only for inline literals, the second is also for code-blocks but can be turned off for inline literals only  via

```rst
latex_elements = {
    'sphinxsetup': 'inlineliteralwraps=false',
}
```

 (cf http://www.sphinx-doc.org/en/master/latex.html#the-sphinx-latex-style-package-options)

which can also be simulated via extra code in LaTeX ``\sphinxsetup{inlineliteralwraps=false}``. (the ``false`` must be lowercased as it is treated by TeX not Python).

This is simplest work-around (which however affects possibility for inline literals to allow line-breaks), but problems will subsist if labels contain one of `` ` < > ' - `` (I recall Sphinx escapes ``<`` and ``>`` in text, I am not sure off-hand what happens if in a target label).

I see in LaTeX writer

    def hyperrefescape(self, ref):
        # type: (unicode) -> unicode
        return self.idescape(ref).replace('-', '\\string-')

A further fix in your specific case would be if also the ``.`` was handled here. But perhaps it would be better for 

    def idescape(self, id):
        # type: (unicode) -> unicode
        return text_type(id).translate(tex_replace_map).\
            encode('ascii', 'backslashreplace').decode('ascii').\
            replace('\\', '_')

to wrap the whole thing in ``\detokenize``.

I am not sure I understand well syntax here, but the idea of the change would be for labels always to be wrapped in ``\detokenize``, be it where they are defined as targets,(*) and where they are used for links. However one should avoid then doing 
something like ```self.idescape(ref).replace('-', '\\string-')``` after this.

(*) in the unlikely case the label definition happens inside a ``\sphinxcode``. A ``\detokenize`` as in ``\label{\detokenize{foo.bar}}`` disappears when the label name is written to ``.aux`` file: it avoids the problems from an active character at the location where `\label` is used, but this protection is only one-shot. In fact one can not use document active characters in ``\label`` in general. But characters made active by (latex) babel work due to a special patching by babel of ``\label/\ref`` LaTeX mechanism.

This would fix issue such as the one you report without modifying behaviour of inline literals for general case.

**Comment by birkenfeld** [comment]
Thanks for the fast diagnosis! Sadly I can't say much about the suggested fixes right now, but making labels "non-active" sounds good.

>How much can the extension customize LaTeX output?

The extension doesn't do anything special with LaTeX, it just puts a `reference` node into a `desc_name` node.

**Comment by jfbu** [comment]
memo: the underlying LaTeX problem is of the same sort as the one from #2026 which was fixed at https://github.com/sphinx-doc/sphinx/commit/53cd2d4f3f386a8dfea28b314112b1a9a882b7c9.

I anticipate the fix of present issue will also fix #2026 but in another way and hence will revert 53cd2d4.

**Comment by jfbu** [comment]
Fixed at 4b42446. (#3206)
